modules/field_event_dispatcher/src/EventSubscriber/Form/FormEntityViewDisplayEditAlterEventSubscriber.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Drupal\field_event_dispatcher\EventSubscriber\Form;
+
+use Drupal\Component\Utility\NestedArray;
+use Drupal\hook_event_dispatcher\HookEventDispatcherInterface;
+use Drupal\core_event_dispatcher\Event\Form\FormIdAlterEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Event subscriber for 'entity_view_display_edit_form' form alter.
+ */
+class FormEntityViewDisplayEditAlterEventSubscriber implements EventSubscriberInterface {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    return [
+      'hook_event_dispatcher.form_entity_view_display_edit_form.alter' => 'formAlter',
+    ];
+  }
+
+  /**
+   * Alter the entity view edit form third-party elements structure.
+   *
+   * This works around the problem that Drupal core nests each invokation of
+   * 'field_formatter_third_party_settings_form' under the machine name of the
+   * module that implements it, which would be the name of this module since and
+   * not the modules that subscribe to our event.
+   *
+   * @param \Drupal\core_event_dispatcher\Event\Form\FormIdAlterEvent $event
+   *   The event object.
+   *
+   * @see \Drupal\field_ui\Form\EntityViewDisplayEditForm::thirdPartySettingsForm()
+   *   This is where Drupal core nests each invokation of
+   *   'field_formatter_third_party_settings_form' under the machine name of the
+   *   module that implements it.
+   *
+   * @see \Drupal\field_event_dispatcher\Event\Field\FieldFormatterThirdPartySettingsFormEvent::addElements()
+   *   This is where event subscribers add their third-party form elements.
+   */
+  public function formAlter(FormIdAlterEvent $event) {
+    $form = &$event->getForm();
+
+    // $form['#fields'] lists all field name keys on this entity.
+    foreach ($form['#fields'] as $fieldName) {
+      // Skip any fields that have no third-party settings.
+      if (!isset($form['fields'][$fieldName]['plugin']['settings_edit_form']['third_party_settings']['field_event_dispatcher'])) {
+        continue;
+      }
+
+      $fieldSettings = &$form['fields'][$fieldName]['plugin']['settings_edit_form'];
+
+      $fieldSettings['third_party_settings'] = NestedArray::mergeDeep(
+        $fieldSettings['third_party_settings'],
+        $fieldSettings['third_party_settings']['field_event_dispatcher']
+      );
```

Comment:
The problem I foresee is that in order to do only one ```NestedArray::mergeDeep()```, we'd have to have it walk over many more nested arrays, since ```$form['fields'][$fieldName]['plugin']['settings_edit_form']['third_party_settings']``` has ```$fieldName``` high up in the keys, which would probably negate any performance gains. What do you think?

Reply:
No I mean your doing it multiple times in a loop now on the `third_party_settings`, it's better to only do it once on all the `third_party_settings` at once. See my example, it builds an array of the special `fieldEventDispatcherSettings` and merge those together in one array merge deep. The `...` will expand the array as arguments in that function. This is based on the fact that 2 times an array_merges is slower then only once with 3 arguments:
```
total = array_merge(a,b)
total = array_merge(total,c)

vs

total = array_merge(a,b,c)

or

combined = [a, b, c];
total = array_merge(...combined)
```

Reply:
I understand what you're saying, but the structure you're suggesting isn't what's in the form array. Maybe this might illustrate it better:

![FieldFormatterThirdPartySettingsFormEvent array dump](https://files.ambientimpact.com/web/drupal/hook_event_dispatcher/FieldFormatterThirdPartySettingsFormEvent.png)

As you can see, each field has its own separate ```third_party_settings```, not a shared one, so we can't merge them all together into one ```third_party_settings```. Each field's ```third_party_settings``` is only being merged once in the code. If you look at what ```$fieldSettings``` contains, it's a reference to the current field's array, not all the combined fields in the form - it's just there as a way to make the code a little more tidy so that the sheer number of nested keys don't get too out of hand.

Also, [```NestedArray```](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Utility%21NestedArray.php/function/NestedArray%3A%3AmergeDeepArray/8.8.x) doesn't use ```array_merge``` internally at all, if that makes any difference.

Reply:
Ah check, I understand you know. Yep your right, this is fine then 👍.

Reply:
Eggcellent. 🎉 Hope I wasn't being too pushy there.

Reply:
Nope, you had a valid point. So no worries 😉.

Reply:
[![cheers with a chinchilla](https://media.giphy.com/media/jOmV27v0xcm51SQ3qj/200w_d.gif)](https://giphy.com/gifs/good-cheers-success-jOmV27v0xcm51SQ3qj)